src/invidious/jsonfilter.cr

```diff
@@ -0,0 +1,251 @@
+      end
+    end
+
+    def self.parse_nest_groups(fields_text : String) : Nil
+      nest_stack = [] of NamedTuple(group_name: String, closing_bracket_index: Int64)
+      bracket_pairs = get_bracket_pairs(fields_text, true)
+
+      text_index = 0
+      regex_index = 0
+
+      while regex_result = self.nest_group_pattern.match(fields_text, regex_index)
+        raw_match = regex_result[0]
+        group_name = regex_result["groupname"]
+
+        text_index = regex_result.begin
+        regex_index = regex_result.end
+
+        if text_index.nil? || regex_index.nil?
+          raise FieldsParser::ParseError.new "Received invalid index while parsing nest groups: text_index: #{text_index} | regex_index: #{regex_index}"
+        end
+
+        offset = raw_match.starts_with?(',') ? 1 : 0
+
+        opening_bracket_index = (text_index + group_name.size) + offset
+        closing_bracket_index = bracket_pairs[opening_bracket_index]
+        content_start = opening_bracket_index + 1
+
+        content = fields_text[content_start...closing_bracket_index]
+
+        if content.empty?
+          raise FieldsParser::ParseError.new "Empty nest group at position #{content_start}"
+        else
+          content = remove_nest_groups(content)
+        end
+
+        while nest_stack.size > 0 && closing_bracket_index > nest_stack[nest_stack.size - 1][:closing_bracket_index]
+          if nest_stack.size
+            nest_stack.pop
+          end
+        end
+
+        group_name.split('/').each do |group_name|
+          nest_stack.push({
+            group_name:            group_name,
+            closing_bracket_index: closing_bracket_index,
+          })
+        end
+
+        if !content.empty?
+          properties = content.split(',')
+
+          properties.each do |prop|
+            nest_list = nest_stack.map { |nest_prop| nest_prop[:group_name] }
+
+            if !prop.empty?
+              if prop.includes?('/')
+                parse_single_nests(prop) { |list| nest_list += list }
+              else
+                nest_list.push prop
+              end
+            else
+              raise FieldsParser::ParseError.new "Empty key in nest list: #{nest_list << prop}"
+            end
+
+            yield nest_list
+          end
+        end
+      end
+    end
+
+    def self.remove_nest_groups(text : String) : String
+      content_bracket_pairs = get_bracket_pairs(text, false)
+
+      content_bracket_pairs.each_key.to_a.reverse.each do |opening_bracket|
+        closing_bracket = content_bracket_pairs[opening_bracket]
+        last_comma = text.rindex(',', opening_bracket) || 0
+
+        text = text[0...last_comma] + text[closing_bracket + 1...text.size]
+      end
+
+      return text.starts_with?(',') ? text[1...text.size] : text
+    end
+
+    def self.get_bracket_pairs(text : String, recursive = true) : BracketIndex
+      istart = [] of Int64
+      d = BracketIndex.new
```

Comment:
Is there a more informative name than `d` you could use here? `bracket_index = BracketIndex.new` maybe?